### ISSUE TITLE
jaeger-zipkin does not bundle libthrift and uses non-shadowed jaeger-thrift library

### ISSUE THREAD

**Issue by mdvorak** [jaegertracing/jaeger-client-java#499]
This is followup to #494.

As title states, `jaeger-zipkin` has different build strategy than `jaeger-thrift`, and actually it might be also broken.
jaeger-thrift shadows private dependencies into internal package to avoid classpath clashes. jaeger-zipkin does not.

This comment states problem with dependency tree - of course it does not work with shadowed library, since jaeger-thrift relocates generated classes in package `com.twitter.zipkin.thriftjava`, and jaeger-zipkin uses them as dependency from jaeger-thrift. But referencing non-shadowed library does not fix the issue, it creates broken build - it depends on classes that are not there, since only shadowed variant is published into maven central.
https://github.com/jaegertracing/jaeger-client-java/blob/a16cd8b8fe669fefcff4abcb46abb3174c06ac8f/jaeger-zipkin/build.gradle#L6

Also it directly references `libthrift` library, which is also relocated inside `jaeger-thrift`, and while this works, classes from `libthrift` exists in two copies in the final build. Behavior should be consistent across whole library.

So, the question is, how to fix this? Using relocated classes from different library is a bit complicated and whole purpose of the relocation and bundling is to provide *private* copies of the dependencies.
But also having duplicate classes across one build is not good solution. 
As I look at it, I suggest dropping `jaeger-zipkin` library altogether for now and moving classes directly into `jaeger-thrift`. Its not many classes, and build will be much easier and reliable.

**Comment by jpkrohling** [comment]
> and actually it might be also broken.

I had the same thought last week (IIRC) and tested it with a brand new application:
https://github.com/jpkrohling/vertx-jaeger-tracer-zipkin-reporter

I'm not claiming it's not broken, but it does work on my tester project.

>  As I look at it, I suggest dropping jaeger-zipkin library altogether for now and moving classes directly into jaeger-thrift. Its not many classes, and build will be much easier and reliable.

I'm not a big fan of this solution. If we do indeed have a problem, I'd like to explore other solutions before we just move Zipkin classes into the Thrift module, especially because we are actually trying to move to the other direction. See https://github.com/jaegertracing/jaeger/issues/925

**Comment by mdvorak** [comment]
I agree with your point of view, but nevertheless, I don't see easy way to achieve that.
Can't be thrift generator tweaked to output `com.twitter` classes into some subpackage directly? So it doesn't have to be relocated?
Another option is not to relocate those, but I don't like it. Its not your package.

**Comment by jpkrohling** [comment]
Unless this library is broken, I wouldn't change it.

**Comment by mdvorak** [comment]
This is dependency tree of project with just `jaeger-zipkin` library:

```
[INFO] com.example:jaeger499:jar:0.0.1-SNAPSHOT
[INFO] \- io.jaegertracing:jaeger-zipkin:jar:0.30.2:compile
[INFO]    +- io.jaegertracing:jaeger-client:jar:0.30.2:compile
[INFO]    |  +- io.jaegertracing:jaeger-thrift:jar:0.30.2:compile
[INFO]    |  +- io.jaegertracing:jaeger-core:jar:0.30.2:compile
[INFO]    |  |  +- io.opentracing:opentracing-api:jar:0.31.0:compile
[INFO]    |  |  +- io.opentracing:opentracing-util:jar:0.31.0:compile
[INFO]    |  |  |  \- io.opentracing:opentracing-noop:jar:0.31.0:compile
[INFO]    |  |  \- com.google.code.gson:gson:jar:2.8.5:compile
[INFO]    |  \- io.jaegertracing:jaeger-tracerresolver:jar:0.30.2:compile
[INFO]    |     \- io.opentracing.contrib:opentracing-tracerresolver:jar:0.1.4:compile
[INFO]    +- io.jaegertracing:jaeger-thrift:jar:no-shadow:0.30.2:compile
[INFO]    +- io.zipkin.reporter2:zipkin-sender-urlconnection:jar:2.7.6:compile
[INFO]    |  +- io.zipkin.reporter2:zipkin-reporter:jar:2.7.6:compile
[INFO]    |  \- io.zipkin.zipkin2:zipkin:jar:2.9.4:compile
[INFO]    \- org.apache.thrift:libthrift:jar:0.11.0:compile
[INFO]       +- org.slf4j:slf4j-api:jar:1.7.25:compile
[INFO]       +- org.apache.httpcomponents:httpclient:jar:4.5.5:compile
[INFO]       |  \- commons-codec:commons-codec:jar:1.11:compile
[INFO]       \- org.apache.httpcomponents:httpcore:jar:4.4.9:compile
```

Notice the `io.jaegertracing:jaeger-thrift:jar:no-shadow:0.30.2:compile` dependency, it is duplicated to `io.jaegertracing:jaeger-thrift:jar:0.30.2:compile`. 
Which is problem - while it contains classes `com.twitter.*` expected by zipkin part, it also contains all classes in `io.jaegertracing`, which are duplicated to shadowed ones. If classloader decides to use versions from `no-shadow`, they are referencing original `okhttp3` classes, not relocated. But missing runtime dependency on those libraries, since it is "compileOnly".

So ignoring the fact you have the classes twice, it works only if classloader loads shadowed classes first. Which is pretty much random. Code will break in runtime, not compile time.
